cmd/contour/servecontext.go

```diff
@@ -232,11 +233,33 @@ func (ctx *serveContext) tlsconfig() *tls.Config {
 		log.Fatalf("unable to append certificate in %s to CA pool", ctx.caFile)
 	}
 
+	// lazily load certificates and key at TLS handshake to ensure that latest
+	// certificates are used in case they have been rotated
 	return &tls.Config{
 		Certificates: []tls.Certificate{cert},
 		ClientAuth:   tls.RequireAndVerifyClientCert,
 		ClientCAs:    certPool,
 		Rand:         rand.Reader,
+		GetConfigForClient: func(*tls.ClientHelloInfo) (*tls.Config, error) {
```

Comment:
Are we sure that we don't want to cache the `tls.Config` for the common case where the certificates have not changed?

Reply:
In theory, cache flushing could be based on `stat()` while being careful to work around how Kubernetes plays with symbolic links.  Alternatively it could be based on hash over the files, which would anyways require reading the file, so why not just return those instead without hashing?

At least from the performance point of view I cannot see reason to implement caching for this use case.   Maybe cache could support scenario where the files were deleted from filesystem but the in-memory versions would keep the system running.  This would not happen in real-life scenario though...

Reply:
The minimum stat information you need is the mtime and the inode number. I expect that this isn't really that performance sensitive, but wanted to make sure that we had the conversation.

Let's link a separate issue to document the issue and consider caching the TLS context. This will give anyone experiencing problems a place to lodge feedback and ensure that we don't forget about it.

Reply:
Ok, yes let's do that!  Could you create this issue, I think you have better idea how to handle this kind of issues that we do not intend to work on, not at least now immediately :)

Reply:
#2228